adoc/config/accordion_toc/extension.rb

```diff
@@ -0,0 +1,188 @@
+          if (ul.style.display === "block") {
+            ul.style.display = "none";
+          } else {
+            ul.style.display = "block";
+          }
+        });
+      }
+
+      toc_top.addEventListener("click", function() {
+        var is_expanded = this.classList.toggle("toc-expanded");
+        for (element of toc_parents) {
+          if (is_expanded) {
+            element.classList.add("toc-expanded");
+            var ul = element.nextElementSibling.nextElementSibling;
+            ul.style.display = "block";
+          }
+          else {
+            element.classList.remove("toc-expanded");
+            var ul = element.nextElementSibling.nextElementSibling;
+            ul.style.display = "none";
+          }
+        }
+      });
+    }
+
+    window.addEventListener("load", addTocClickHandlers);
+</script>
+'
+
+  # Postprocess the HTML performing the following modifications:
+  #
+  # * Each heading in the TOC is represented by an <li> element.  If the heading
+  #   has sub-headings, the <li> is follows by a <ul>.  Find these <li> elements
+  #   that have sub-headings and add a <span class="toc-parent"> after the <li>.
+  #   The script above uses the class name to attach a click handler.  The CSS
+  #   style sheet also uses the class name to attach an icon that the user can
+  #   click.
+  #
+  # * Keep track of the heading levels and modify the <li> and <ul> elements to
+  #   initially expand the N topmost levels according to the attribute
+  #   "toclevels-expanded".  The <li> element for an expanded level is given the
+  #   class name "toc-expanded".  The CSS style sheet uses this to change the
+  #   icon to indicate that the TOC level is expanded.  For an unexpanded
+  #   element, we hide the <ul> that follows by adding 'style="display:none;"'.
+  #
+  # * Add a <span id="toc-top"> element to the TOC title line.  The script above
+  #   uses the ID to attach a click handler, and the CSS style sheet uses the ID
+  #   to attach a clickable icon.
+  #
+  # The Asciidoctor postprocessor pass just gets a big string of HTML, not a DOM
+  # tree.  Rather then trying to parse the HTML, we do some fairly simplistic
+  # pattern matching to find the relevant HTML elements.  This pattern matching
+  # relies on the current output format of the Asciidoctor HTML generator.  If
+  # that format changes in the future, we will either need to change the
+  # pattern matching in this script or do something more robust by really
+  # parsing the HTML.
+  def process document, output
+
+    if document.basebackend? 'html'
+      if document.attr? 'toclevels-expanded'
+        toc_levels_expanded = (document.attr 'toclevels-expanded').to_i
+      else
+        toc_levels_expanded = -1
+      end
+      new_output = ''
+      is_in_toc = false
+      hide_next_ul = false
+      toc_level = 0
+      lines = output.lines
+      num_lines = lines.length-1
+      for i in 0..num_lines
```

Comment:
`new_output = outputs.map{ |line| }` is the ruby way (or event better just ` outputs.map{ |line| }` , as the implicit return of the map is what you are doing

Reply:
Can I really use `map` here?  Note that the algorithm needs a 1-line lookahead.  Can you do that easily with `map`?

Reply:
Yes, it's called `each_cons`  so ouputs.map.each_cons(2) { |line, new_line| }

Reply:
Thanks: d513921589b2f96cb05c3fb1731267a9be522f56.

The loop never cares about the last line of input, so we can do without the `+[""]`.